pkg/audit/manager.go

```diff
@@ -0,0 +1,360 @@
+			Kind:      ar.rkind,
+			Name:      ar.rname,
+			Namespace: ar.rnamespace,
+			Message:   ar.message,
+		})
+	}
+	raw, err := json.Marshal(statusViolations)
+	if err != nil {
+		return err
+	}
+	// need to convert to []interface{}
+	violations := make([]interface{}, 0)
+	err = json.Unmarshal(raw, &violations)
+	if err != nil {
+		return err
+	}
+	// update constraint status auditTimestamp
+	unstructured.SetNestedField(instance.Object, timestamp, "status", "auditTimestamp")
+	// update constraint status violations
+	if len(violations) == 0 {
+		_, found, err := unstructured.NestedSlice(instance.Object, "status", "violations")
+		if err != nil {
+			return err
+		}
+		if found {
+			unstructured.RemoveNestedField(instance.Object, "status", "violations")
+			log.Info("removed status violations", "constraintName", constraintName)
+		}
+		err = ucloop.client.Update(ctx, instance)
+		if err != nil {
+			return err
+		}
+	} else {
+		unstructured.SetNestedSlice(instance.Object, violations, "status", "violations")
+		log.Info("update constraint", "object", instance)
+		err = ucloop.client.Update(ctx, instance)
+		if err != nil {
+			return err
+		}
+		log.Info("updated constraint status violations", "constraintName", constraintName, "count", len(violations))
+	}
+	return nil
+}
+
+func truncateString(str string, size int) string {
+	shortenStr := str
+	if len(str) > size {
+		if size > 3 {
+			size -= 3
+		}
+		shortenStr = str[0:size] + "..."
+	}
+	return shortenStr
+}
+
+type updateConstraintLoop struct {
+	uc      map[string]unstructured.Unstructured
+	client  client.Client
+	stop    chan struct{}
+	stopped chan struct{}
+	ul      map[string][]auditResult
+	ts      string
+}
+
+func (ucloop *updateConstraintLoop) update() {
+	defer close(ucloop.stopped)
+	updateLoop := func() (bool, error) {
+		for _, item := range ucloop.uc {
+			select {
+			case <-ucloop.stop:
+				return true, nil
+			default:
+				failure := false
+				// if constraint is in updatedLists, then update its violations
+				if constraintAuditResults, ok := ucloop.ul[item.GetSelfLink()]; !ok {
+					err := ucloop.updateConstraintStatus(context.Background(), &item, emptyAuditResults, ucloop.ts)
+					if err != nil {
+						failure = true
+						log.Error(err, "could not update constraint status", "name", item.GetName(), "namespace", item.GetNamespace())
+					}
+				} else {
+					// update the constraint
+					err := ucloop.updateConstraintStatus(context.Background(), &item, constraintAuditResults, ucloop.ts)
+					if err != nil {
+						failure = true
+						log.Error(err, "could not update constraint status", "name", item.GetName(), "namespace", item.GetNamespace())
+					}
+				}
+				if !failure {
+					delete(ucloop.uc, item.GetSelfLink())
+				}
+			}
+		}
+		if len(ucloop.uc) == 0 {
+			return true, nil
+		}
+		return false, nil
+	}
+
+	if err := wait.ExponentialBackoff(wait.Backoff{
+		Duration: 5 * time.Second,
```

Comment:
I think there's some tuning to be done here. For my config controller I'm not sold on the 5 sec initial retry as failure is somewhat expected. We will see what the practical effects are once we have more in-the-wild testing.

For audit I think we can make some reasonable assumptions based off the audit tick intervals (looks like default is 60 seconds)

Currently a full backoff cycle of 5 iterations will have its last retry be anywhere between 160 seconds and 320 seconds from the first (5s * (2 ** 5 - 1) * [1 - 2 jitter])

Because anything above 60 seconds is moot, we could drop the base duration to 1 second, which yields a spread of 32 - 64 seconds at the maximum

Reply:
Changed `Duration` to 1 second.